rescript

```diff
@@ -318,251 +197,382 @@ if (
-      // On other systems, we just have to assume that any change is valid.
-      // This could cause problems if source builds (generating js files in the same directory) are supported.
-      if (!fileName) return true;
-
-      return !(
-        fileName === ".merlin" ||
-        fileName.endsWith(".js") ||
-        fileName.endsWith(".mjs") ||
-        fileName.endsWith(".cjs") ||
-        fileName.endsWith(genTypeFileExtension) ||
-        watchGenerated.indexOf(fileName) >= 0 ||
-        fileName.endsWith(".swp")
-      );
+    var portNumber;
+    if (hostAndPortNumber.length === 1) {
+      portNumber = parseInt(hostAndPortNumber[0]);
+    } else {
+      webSocketHost = hostAndPortNumber[0];
+      portNumber = parseInt(hostAndPortNumber[1]);
     }
-
-    /**
-     * @return {boolean}
-     */
-    function needRebuild() {
-      return reasonsToRebuild.length !== 0;
+    if (!isNaN(portNumber)) {
+      webSocketPort = portNumber;
     }
+    withWebSocket = true;
+    dlog(`WebSocket host & port number: ${webSocketHost}:${webSocketPort}`);
+  }
 
-    function logFinish(code) {
-      let log = `>>>> Finish compiling`;
-      if (code !== 0) {
-        log = log + " (exit: " + code + ")";
+  const rescriptWatchBuildArgs = verbose
+    ? ["build", "-no-deps", "-verbose"]
+    : ["build", "-no-deps"];
+
+  function notifyClients() {
+    wsClients = wsClients.filter(x => !x.closed && !x.socket.destroyed);
+    var wsClientsLen = wsClients.length;
+    dlog(`Alive sockets number: ${wsClientsLen}`);
+    var data = '{"LAST_SUCCESS_BUILD_STAMP":' + LAST_SUCCESS_BUILD_STAMP + "}";
+    for (var i = 0; i < wsClientsLen; ++i) {
+      // in reverse order, the last pushed get notified earlier
+      var client = wsClients[wsClientsLen - i - 1];
+      if (!client.closed) {
+        client.sendText(data);
       }
-      if (std_is_tty) {
-        log = "\x1b[36m" + log + "\x1b[0m";
+    }
+  }
+
+  function setUpWebSocket() {
+    var WebSocket = require("./lib/minisocket.js").MiniWebSocket;
+    var id = setInterval(notifyClients, 3000);
+    require("http")
+      .createServer()
+      .on("upgrade", function (req, socket, upgradeHead) {
+        dlog("connection opened");
+        var ws = new WebSocket(req, socket, upgradeHead);
+        socket.on("error", function (err) {
+          dlog(`Socket Error ${err}`);
+        });
+        wsClients.push(ws);
+      })
+      .on("error", function (err) {
+        // @ts-ignore
+        if (err !== undefined && err.code === "EADDRINUSE") {
+          var error = isTtyStd ? `\x1b[1;31mERROR:\x1b[0m` : `ERROR:`;
+          console.error(`${error} The websocket port number ${webSocketPort} is in use.
+Please pick a different one using the \`-ws [host:]port\` flag from bsb.`);
+        } else {
+          console.error(err);
+        }
+        process.exit(2);
+      })
+      .listen(webSocketPort, webSocketHost);
+  }
+
+  function watchBuild(watchConfig) {
```

Comment:
This function is not touched. Only the `watchGenerated` definition moved to a higher scope.